File: Lecture41/reptation.py
import math
import random
import sys
import timeit
from collections import deque

def create_snake(steps, config='coil'):
    global snake
    snake = deque()         # double-ended priority queue

    # initialize with head and tail at x=0, y=0
    x, y = [0, 0]
    snake.append([x, y])

    # choice of 'coil', 'stair', or straight line
    if config == 'coil':
        # tight coil East, North, West^2, South^2, E^3, N^3, ...
        directions = [ ['East', 'North'], ['West', 'South'] ]
        ds = directions[0]  # step E then N
        n_ds = 1            # number of directed steps in each direction
        step = 0            # number of step done
        while step < steps:
            for i in range(2):
                for n in range(n_ds):
                    if ds[i] == 'East':
                        x += 1
                    elif ds[i] == 'North':
                        y += 1
                    elif ds[i] == 'West':
                        x -= 1
                    else:
                        y -= 1
                    if step < steps:
                        snake.append([x, y])
                    step += 1
            n_ds += 1       # increase number of directed steps
    elif config == 'stair':
        # step randomly East or North
        for step in range(steps):
            if random.randrange(2) == 0:
                x += 1
            else:
                y += 1
            snake.append([x, y])
    else:
        # straight line East
        for step in range(steps):
            x += 1
            snake.append([x,y])

def random_allowed(head, neck):
    # given x,y coordinates of head and neck
    # return a random site for the head to move to

    # construct a list of 4 nearest neighbors of the head and remove the neck
    x, y = head
    neighbors = [ [x + 1, y], [x - 1, y], [x, y + 1], [x, y -1] ]
    neighbors.remove(neck)

    # return a random allowed site
    return neighbors[random.randrange(3)]

def reptate_succeeded():    # attempt random move and return true if succeeded
    global snake

    # choose one end of the chain at random to be the head by tossing a coin
    if random.randrange(2) == 0:
        snake.reverse()     # switch head and tail

    # choose a random head move
    next = random_allowed(snake[0], snake[1])

    # check whether it is an interior site other than the neck
    tail = snake[-1]
    if next in snake and next != tail:
        return False        # move failed, retain current configuration

    # repate by removing the tail and advancing the head
    snake.remove(tail)
    snake.appendleft(next)
    return True

print(" Reptation Method for Self-Avoiding Walks on a Square Lattice")
print(" ------------------------------------------------------------")
n_steps = int(input(" Enter maximum number of steps in walk: "))
n_walks = int(input(" Enter maximum number walks to generate: "))
config = input(" Enter initial configuration (coil, stair, line): ")

file = open("reptation.data", "w")
print(" Steps   <r^2>       Std. Dev.   Success%  CPU secs")

for steps in range(1, n_steps + 1):
    r2_sum = 0.0
    r4_sum = 0.0
    success = 0
    start_time = timeit.default_timer()

    create_snake(steps, config)
    for i in range(n_walks):
        if reptate_succeeded():
            success += 1
            head = snake[0]
            tail = snake[-1]
            r2 = (head[0] - tail[0])**2 + (head[1] - tail[1])**2
            r2_sum += r2
            r4_sum += r2**2

    end_time = timeit.default_timer()
    r2_av = r2_sum / n_walks
    std_dev = math.sqrt(r4_sum / n_walks - r2_av**2)
    success_percent = success / float(n_walks) * 100.0
    cpu = (end_time - start_time)

    result = ('  '
            + str(steps).ljust(4) + '   '
            + str(r2_av)[:10].ljust(10) + '  '
            + str(std_dev)[:10].ljust(10) + '  '
            + str(success_percent)[:8].ljust(8) + '  '
            + str(cpu))
    print(result)
    file.write(result + '\n')

file.close()
print(" Data in file reptation.data")

File: Lecture41/reptation.cpp
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <deque>
#include <iomanip>
#include <iostream>
#include <fstream>
#include <set>
using namespace std;

inline double std_rand()
{
    return rand() / (RAND_MAX + 1.0);
}

struct Site {               // object to represent lattice site
    int x;                  // x coordinate
    int y;                  // y coordinate

    // strict weak-ordering comparison operator required by set
    bool operator< (const Site& s) const {
        const int maxDiff = 9999;   // must be larger than x or y difference
        return (x - s.x) + maxDiff * (y - s.y) < 0;
    }
};

deque<Site> snake;          // double-headed reptile
set<Site> occupiedSites;    // set of occupied sites

bool occupied(Site s) {     // return true if s is occupied
    return occupiedSites.find(s) != occupiedSites.end();
}

void clear() {              // remove all sites
    snake.clear();
    occupiedSites.clear();
}

void addBack(Site s) {      // add s to back of reptile
    snake.push_back(s);
    occupiedSites.insert(s);
}

void addFront(Site s) {     // add s to back of reptile
    snake.push_front(s);
    occupiedSites.insert(s);
}

void removeBack() {         // remove back end of reptile
    occupiedSites.erase(snake.back());
    snake.pop_back();
}

void removeFront() {        // remove front end of reptile
    occupiedSites.erase(snake.front());
    snake.pop_front();
}

const int EAST = 0, NORTH = 1, WEST = 2, SOUTH = 3, DIRECTIONS = 4;

const int                   // initial configuration choices
    STAIR = 1,              // East-North random staircase
    COIL  = 2,              // tight coil East, North, West, West, South, ...
    LINE  = 3;              // straight line East

void createSnake(           // make a snake with
    int steps,              // this number of segments
    int config = LINE)      // and this initial configuration
{
    clear();                // remove all sites

    Site s;
    s.x = s.y = 0;
    addFront(s);

    for (int step = 1; step <= steps; step++) {

        int stp = 0, dir = EAST;    // initialize variables to construct coil

        switch (config) {
        case STAIR:             // add randomly East or North
            std_rand() < 0.5 ? ++s.x : ++s.y;
            break;
        case COIL:              // add in sequence E,N,W,W,S,S,E,E,E,N,N,N,...
            while (stp < step)
                stp += ++dir / 2;
                switch ((dir + 2) % DIRECTIONS) {
                case EAST  :  ++s.x;  break;
                case NORTH :  ++s.y;  break;
                case WEST  :  --s.x;  break;
                case SOUTH :  --s.y;  break;
                }
                break;
        case LINE:              // add East
        default:                // also the default
            ++s.x;
        }

        addFront(s);
    }
}

Site randomAllowed(         // return a random allowed site
    Site head,              // adjacent to this head site
    Site neck)              // excluding this neck site
{
    deque<Site> allowed;

    // find the 3 allowed directions and add site to allowed deque

    for (int direction = EAST; direction < DIRECTIONS; direction++) {

        Site s;
        s.x = head.x;
        s.y = head.y;

        switch (direction) {
        case EAST:
            if ( !(head.x == neck.x - 1 && head.y == neck.y) ) {
                ++s.x;
                allowed.push_back(s);
            }
            break;
        case NORTH:
            if ( !(head.x == neck.x && head.y == neck.y - 1) ) {
                ++s.y;
                allowed.push_back(s);
            }
            break;
        case WEST:
            if ( !(head.x == neck.x + 1 && head.y == neck.y) ) {
                --s.x;
                allowed.push_back(s);
            }
            break;
        case SOUTH:
            if ( !(head.x == neck.x && head.y == neck.y + 1) ) {
                --s.y;
                allowed.push_back(s);
            }
            break;
        }
    }

    // choose and return a random allowed site
    return allowed[int(2.999999 * std_rand())];
}

bool reptate() {            // attempt random move and return true if succeeded

    if (snake.size() < 2)               // cannot reptate
        return false;

    Site head, neck, sNext;

    if (std_rand() < 0.5) {    // choose front end of snake
        head = snake[0];
        neck = snake[1];
        sNext = randomAllowed(head, neck);
        if (occupied(sNext))
            return false;
        removeBack();
        addFront(sNext);
    } else {                            // choose back end of snake
        int n = snake.size();
        head = snake[n - 1];
        neck = snake[n - 2];
        sNext = randomAllowed(head, neck);
        if (occupied(sNext))
            return false;
        removeFront();
        addBack(sNext);
    }

    return true;
}

double rSquared() {         // end-to-end size squared
    if (snake.size() < 2)
        return 0.0;
    double dx = snake.front().x - snake.back().x;
    double dy = snake.front().y - snake.back().y;
    return dx * dx + dy * dy;
}

int main() {

    cout << " Reptation Method for Self-Avoiding Walks on a Square Lattice\n"
         << " ------------------------------------------------------------\n"
         << " Enter maximum number of steps in walk: ";
    int n_steps, n_walks;
    cin >> n_steps;
    cout << " Enter number of walks to generate: ";
    cin >> n_walks;
    cout << " Enter initial configuration 1 = stair, 2 = coil, 3 = line: ";
    int config;
    cin >> config;

    ofstream file("reptation.data");
    cout << " Steps   <r^2>       Std. Dev.   Success%  CPU secs" << endl;

    for (int steps = 1; steps <= n_steps; steps++) {

        double r2sum = 0;
        double r4sum = 0;
        int success = 0;
        clock_t startTime = clock();

        createSnake(steps, config);

        for (int i = 0; i < n_walks; i++) {
            if (reptate())
                ++success;
            double r2 = rSquared();
            r2sum += r2;
            r4sum += r2 * r2;
        }

        clock_t endTime = clock();

        double r2av = r2sum / n_walks;
        double stdDev = sqrt(r4sum / n_walks - r2av * r2av);
        double successPercent = success / double(n_walks) * 100.0;
        double cpu = (endTime - startTime) / double(CLOCKS_PER_SEC);

        cout << right << setw(4)  << steps           << "     "
             << left  << setw(10) << r2av            << "  "
             << left  << setw(10) << stdDev          << "  "
             << left  << setw(8)  << successPercent  << "  "
            << left  << setw(8)  << cpu             << '\n';
        file << steps << '\t' << r2av << '\t' << stdDev
             << '\t' << successPercent << '\t' << cpu << '\n';
    }

    file.close();
    cout << " Data in file reptation.data" << endl;
}
